scripts/get_orphaned_functions.py

```diff
@@ -0,0 +1,243 @@
+
+    function_names = sorted(set(function_names))
+
+    # Print the function names we extracted and their count
+    # print(color_codes[13] + "Function names:", function_names, "\n", ENDC)
+    # print(color_codes[2] + "Length of function List", len(function_names), "\n", ENDC)
+
+    return function_names
+
+
+# Define a function to check if a function is utilized within its module
+def is_function_utilized_within_module(function_name, module_r_path):
+    """
+    Check if the function is utilized within its module.
+
+    Args:
+    function_name (str): The name of the function to check.
+    module_r_path (str): The path to the module's R directory.
+
+    Returns:
+    bool: True if the function is utilized within the module, False otherwise.
+    """
+
+    # Regex pattern to detect usage of the function
+    function_pattern = re.compile(r"\b" + re.escape(function_name) + r"\b\s*\(")
+
+    # Traverse all files in the given module R directory
+    for r_file_name in os.listdir(module_r_path):
+        # Check if the file is an R file
+        if r_file_name.endswith(".R"):
+            file_path = os.path.join(module_r_path, r_file_name)
+            with open(file_path, "r") as file:
+                file_content = file.read()
+
+                # Checks if the function is utilized in the file by searching for the function name
+                if function_pattern.search(file_content):
+                    # print(f'{color_codes[2]}Function "{function_name}" is utilized in file: {file_path}{ENDC}')
+                    return True
+
+    # If no file uses the function
+    # print(f'{color_codes[1]}Function "{function_name}" is not utilized in module: {module_r_path}, {ENDC}')
+    return False
+
+
+def is_function_utilized_externally(function_name, module_name):
+    """
+    Check if the function from a specific module is utilized in any other module.
+
+    Args:
+    function_name (str): The name of the function to check.
+    module_name (str): The name of the module where the function is originally defined.
+
+    Returns:
+    bool: True if the function is utilized externally, False otherwise.
+    """
+    function_pattern = re.compile(r"\b" + re.escape(function_name) + r"\b")
+    func_location_module = os.path.join(main_dir, "modules", module_name)
+    search_directories = [
+        os.path.join(main_dir, "modules"),
+        os.path.join(main_dir, "models"),
+        os.path.join(main_dir, "scripts"),
+        os.path.join(main_dir, "base"),
+        os.path.join(main_dir, "tests"),
+        os.path.join(main_dir, "base"),
+    ]
+
+    # print(color_codes[2] + "Function name:",function_name," \t Function's Location modules :",func_location_module,ENDC,)
+
+    for directory in search_directories:
+        for root, dirs, files in os.walk(directory):
+            if (
+                root.__contains__(func_location_module)
+                or root.__contains__("/man")
+                or root.__contains__("/vignettes")
```

Comment:
I would count a usage in another package's vignette as an external usage.

Reply:
Thanks for the heads up. Removed `vignettes` from excluded directory